src/relax/transform/to_mixed_precision.cc

```diff
@@ -0,0 +1,479 @@
+ public:
+  explicit ToMixedPrecisionRewriter(const VarDTypeMap* var_dtype_map, DataType output_dtype)
+      : var_dtype_map_(var_dtype_map), output_dtype_(output_dtype) {}
+
+ private:
+  using VarMap = std::unordered_map<Var, Var, ObjectPtrHash, ObjectPtrEqual>;
+
+  // TODO(@bohan): implements some postorder function accepts a visitor closure
+  class VarReplacer : public ExprMutator {
+   public:
+    explicit VarReplacer(const ToMixedPrecisionRewriter::VarMap& realized) : realized_(realized) {}
+
+    static Expr Replace(const Expr& expr, const ToMixedPrecisionRewriter::VarMap& realized) {
+      VarReplacer replacer(realized);
+      return replacer(expr);
+    }
+
+   private:
+    Expr VisitExpr_(const VarNode* op) final {
+      Var var = GetRef<Var>(op);
+      auto it = realized_.find(var);
+      return it == realized_.end() ? var : it->second;
+    }
+
+    const ToMixedPrecisionRewriter::VarMap& realized_;
+  };
+
+  // Util function to rewrite the expr to the given dtype
+  // we first replace the vars with the realized vars
+  // then rewrite each leaf tensor to the given dtype if necessary
+  // Note that this function only accepts expr with nested tensor type
+  Expr RewriteExpr(const Expr& expr, const NDType& to) {
+    auto fvisitleaf = [&](const Expr& expr, const NDType& to) -> Expr {
+      const auto* tensor = GetStructInfoAs<TensorStructInfoNode>(expr);
+      ICHECK(tensor != nullptr) << "Only support rewriting tensor expr";
+      if (NDTypeEqual()(to, NDTypeFrom(expr))) return expr;
+      return astype(expr, DataType(String2DLDataType(to.LeafValue())));
+    };
+    return TransformTupleLeaf<String>(VarReplacer::Replace(expr, realized_var_), to, fvisitleaf);
+  }
+
+  Array<Expr> RewriteArgs(const Array<Expr>& args, const Array<NDType>& to) {
+    ICHECK_EQ(args.size(), to.size()) << "Inconsistent args and to";
+    Array<Expr> new_args;
+    for (size_t i = 0; i < args.size(); ++i) {
+      new_args.push_back(RewriteExpr(args[i], to[i]));
+    }
+    return new_args;
+  }
+
+  Array<Expr> RewriteArgs(const Array<Expr>& args, DataType to) {
+    Array<Expr> new_args;
+    for (const Expr& arg : args) {
+      if (IsNestedTensor(arg)) {
+        new_args.push_back(RewriteExpr(arg, NDTypeFrom(arg, to)));
+      } else {
+        new_args.push_back(arg);
+      }
+    }
+    return new_args;
+  }
+
+  void RealizeVarDef(const Var& var, const Var& cur_realize) {
+    // realize the var to the required dtype if necessary
+    auto it = var_dtype_map_->find(var);
+    if (it == var_dtype_map_->end()) return;
+    // Currently var is realized to cur_realize
+    realized_var_[var] = cur_realize;
+    // Rewrite cur_realize to the required dtype
+    Expr rewrite = RewriteExpr(cur_realize, it->second);
+    // If cur_realize is not rewritten, we don't need to emit a new var
+    if (!rewrite.same_as(cur_realize)) {
+      // Emit a new var, and update the realized var map
+      realized_var_[var] = builder_->Emit(rewrite);
```

Comment:
Is there any reason we set the mapping twice? It seems to me that this RewriteExpr won’t inquiry the realized version of `var` here.